### PR TITLE
Raise minimum C++ standard to C++20

### PR DESCRIPTION
**Pull request by nullsystem**
* GCC compiler error fixes
    * Comparison is ambiguous - EX: `m_SymbolTable.Find( pString ) != UTL_INVAL_SYMBOL;` turned to `m_SymbolTable.Find( pString ).IsValid();` or adding `()` to use `operator()` on the few that are pointer comparions
    * `concept` is now a reserved keyword and there's a lot of variables in the SDK that uses the name, just rename them to `aiconcept` (as they're typically `AIConcept_t` types) and that fixes it
* GCC compiler warning fixes
    * Some additional warning fixes around `volatile` `++/--` depreciation, no more `register`, and warning version of ambiguous issue
* MSVC compiler error fixes (after GCC fixes applied)
    * `const` especially often around `const char *` is now stricter, must explicitly cast between non-const and const
    * For the time being, some of them are proper fix, others, just cast to mute them. Not like it's worse than it already was anyway.
* Checked and run (after both applied):
    - [x] MSVC Debug
    - [x] MSVC Release
    - [x] GCC (SteamRT + Native) Debug
    - [x] GCC (SteamRT + Native) Release
* Compiler versions checked on:
    * SteamRT: `g++ (SteamRT 10.3.0-3+steamrt3.1+bsrt3.1) 10.3.0`
    * Native: `g++ (GCC) 14.1.1 20240522`
* The server instructions need to be updated for Linux as with C++20 it'll only work with the `LD_LIBRARY_PATH` to steam-runtime output trick then run `srcds_linux` directly

### CODE REVIEW

File: mp/src/public/tier0/threadtools.h
Comment (Rainyan) [inline mp/src/public/tier0/threadtools.h:1]: For Qt Creator 6.0.2, `g++ (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0` build on Linux, I'm getting:

```
/rebuild/neo/mp/src/public/tier0/threadtools.h:869: error: expected unqualified-id before ‘const’
In file included from /rebuild/neo/mp/src/public/mathlib/vector.h:31,
                 from /rebuild/neo/mp/src/public/mathlib/mathlib.h:13,
                 from /rebuild/neo/mp/src/public/tier1/bitbuf.h:19,
                 from /rebuild/neo/mp/src/tier1/bitbuf.cpp:9:
/rebuild/neo/mp/src/public/tier0/threadtools.h:869:33: error: expected unqualified-id before ‘const’
  869 |         CAutoLockT<MUTEX_TYPE>( const CAutoLockT<MUTEX_TYPE> & );
      |                                 ^~~~~
```
and:
```
/rebuild/neo/mp/src/public/tier0/threadtools.h:869: error: expected ‘)’ before ‘const’
/rebuild/neo/mp/src/public/tier0/threadtools.h:869:32: error: expected ‘)’ before ‘const’
  869 |         CAutoLockT<MUTEX_TYPE>( const CAutoLockT<MUTEX_TYPE> & );
      |                               ~^~~~~~
      |                                )
cc1plus: note: unrecognized command-line option ‘-Wno-template-id-cdtor’ may have been intended to silence earlier diagnostics
```

Comment (nullsystem) [reply mp/src/public/tier0/threadtools.h:1]: @Rainyan Try changing the line to: `CAutoLockT( const CAutoLockT<MUTEX_TYPE> & );` and see if it worked out. SteamRT and native `g++ (GCC) 14.1.1 20240522` didn't pick up on this for me.

**Comment by Rainyan** [comment]
For the file `src/public/tier1/datamanager.h` on Linux, I'm seeing:

```
/rebuild/neo/mp/src/public/tier1/datamanager.h:136: error: expected unqualified-id before ‘unsigned’
In file included from /rebuild/neo/mp/src/tier1/datamanager.cpp:10:
/rebuild/neo/mp/src/public/tier1/datamanager.h:136:75: error: expected unqualified-id before ‘unsigned’
  136 |         CDataManager<STORAGE_TYPE, CREATE_PARAMS, LOCK_TYPE, MUTEX_TYPE>( unsigned int size = (unsigned)-1 ) : BaseClass(size) {}
      |                                                                           ^~~~~~~~
```
```
/rebuild/neo/mp/src/public/tier1/datamanager.h:136: error: expected ‘)’ before ‘unsigned’
/rebuild/neo/mp/src/public/tier1/datamanager.h:136:70: error: expected ‘)’ before ‘unsigned’
  136 |     CDataManager<STORAGE_TYPE, CREATE_PARAMS, LOCK_TYPE, MUTEX_TYPE>( unsigned int size = (unsigned)-1 ) : BaseClass(size) {}
      |                                                                     ~^~~~~~~~~
      |                                                                      )
```

```
/rebuild/neo/mp/src/public/tier1/datamanager.h:139: error: template-id not allowed for destructor
/rebuild/neo/mp/src/public/tier1/datamanager.h:139:9: error: template-id not allowed for destructor
  139 |         ~CDataManager<STORAGE_TYPE, CREATE_PARAMS, LOCK_TYPE, MUTEX_TYPE>()
      |         ^
cc1plus: note: unrecognized command-line option ‘-Wno-template-id-cdtor’ may have been intended to silence earlier diagnostics
```

```
/rebuild/neo/mp/src/public/tier1/datamanager.h:254: error: cast from pointer to smaller type 'unsigned int' loses information
```

**Comment by Rainyan** [comment]
> @Rainyan Updated the branch to attempt to fix those and previous: [b145e7b](https://github.com/NeotokyoRebuild/neo/commit/b145e7b289f3fbeaccae2b284e3536aee972985c)

Thanks, it's building correctly now :+1:

> I've also added a commented out flag to cmake as a possible solution that might also workaround the 2nd part also so try that out if it maybe needed.

I tried building both with and without the CMakeLists `ms-extensions` flag, and both seemed to work. I didn't see any difference in the amount of warnings emitted either, so perhaps we'd wanna keep this one commented out by default?

**Review by Rainyan — APPROVED** [review]

**Comment by nullsystem** [comment]
Fixed typo and changed README/CONTRIBUTION.md to fit with C++20.

> I tried building both with and without the CMakeLists `ms-extensions` flag, and both seemed to work. I didn't see any difference in the amount of warnings emitted either, so perhaps we'd wanna keep this one commented out by default?

@Rainyan Yeah I'll just keep it commented out for now.

File: README.md [hunk 1]
```diff
@@ -170,20 +170,22 @@ You may or may not need this, but if NT;RE crashes/segfaults on launch, then ins
 and set: `LD_PRELOAD=/usr/lib32/libtcmalloc.so %command%` as the launch argument.
 
 ## Server instructions
-1. To run a server, install "Source SDK Base 2013 Dedicated Server".
+1. To run a server, install "Source SDK Base 2013 Dedicated Server" (appid 243750).
```
Comment (Rainyan) [inline README.md:173]: 243750 is Source SDK Base 2013 Multiplayer according to steamdb, did you mean to have 244310 here?

Comment (nullsystem) [reply README.md:173]: whoops good catch, yeah I mean 244310

**Review by Rainyan — APPROVED** [review]

**Comment by Masterkatze** [comment]
LGTM

**Pull request merged**